### ISSUE TITLE
Operator v0.64.0 got panic after creating the CRD PrometheusAgent

### ISSUE THREAD

**Issue by CheyiLin** [prometheus-operator/prometheus-operator#5508]
**What happened?**

```
panic: interface conversion: interface {} is *v1alpha1.PrometheusAgent, not *v1.Prometheus                                                                                                                                                     
                                                                                                                                                                                                                                               
goroutine 678 [running]:                                                                                                                                                                                                                       
github.com/prometheus-operator/prometheus-operator/pkg/prometheus.(*Collector).Collect(0xc0007bd548, 0xc000751768?)                                                                                                                            
    /home/runner/work/prometheus-operator/prometheus-operator/pkg/prometheus/collector.go:69 +0xfc                                                                                                                                             
github.com/prometheus/client_golang/prometheus.(*wrappingCollector).Collect.func1()                                                                                                                                                            
    /home/runner/go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/wrap.go:131 +0x2e                                                                                                                                           
created by github.com/prometheus/client_golang/prometheus.(*wrappingCollector).Collect                                                                                                                                                         
    /home/runner/go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/wrap.go:130 +0x8d
```

**Did you expect to see something different?**

No panic.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create Prometheus agent by applying CRD `PrometheusAgent`.
2. Check the provisioned statefulset `prom-agent-<name>`.
3. A few seconds later, the `prometheus-operator` crashed.

**Environment**

* Prometheus Operator version:

`quay.io/prometheus-operator/prometheus-operator:v0.64.0`

* Kubernetes version information:

```
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.6", GitCommit:"ff2c119726cc1f8926fb0585c74b25921e866a28", GitTreeState:"clean", BuildDate:"2023-01-18T19:22:09Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"darwin/arm64"}
Server Version: version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.6-eks-48e63af", GitCommit:"9f22d4ae876173884749c0701f01340879ab3f95", GitTreeState:"clean", BuildDate:"2023-01-24T19:19:02Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster kind:

AWS EKS with EKS managed node groups.

* Manifests:

- CRD `PrometheusAgent`:

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: PrometheusAgent
metadata:
  name: test
  namespace: monitoring
spec:
  paused: false
  serviceAccountName: prometheus
  replicas: 1
  version: v2.43.0
  enableFeatures: []
  storage: {}
  resources:
    limits:
      memory: 1000Mi
    requests:
      cpu: 200m
  securityContext:
    fsGroup: 2000
    runAsNonRoot: true
    runAsUser: 1000
  storage:
    ephemeral:
      volumeClaimTemplate:
        spec:
          accessModes:
          - ReadWriteOnce
          storageClassName: gp3
          resources:
            requests:
              storage: 1Gi
  scrapeInterval: 30s
  scrapeTimeout: 10s
  replicaExternalLabelName: __replica__
  externalLabels:
    cluster: my-cluster
  remoteWrite: []
  serviceMonitorSelector: {}
  serviceMonitorNamespaceSelector: {}
  podMonitorSelector: {}
  podMonitorNamespaceSelector: {}
  probeSelector: {}
  probeNamespaceSelector: {}
```

- `prometheus-operator` args:

```
--log-level=debug                                                                                                                                                                                                                        
--kubelet-service=kube-system/kube-prometheus-stack-kubelet                                                                                                                                                                              
--localhost=127.0.0.1                                                                                                                                                                                                                    
--prometheus-config-reloader=quay.io/prometheus-operator/prometheus-config-reloader:v0.64.0                                                                                                                                              
--config-reloader-cpu-request=200m                                                                                                                                                                                                       
--config-reloader-cpu-limit=200m                                                                                                                                                                                                         
--config-reloader-memory-request=50Mi                                                                                                                                                                                                    
--config-reloader-memory-limit=50Mi                                                                                                                                                                                                      
--thanos-default-base-image=quay.io/thanos/thanos:v0.30.2                                                                                                                                                                                
--web.enable-tls=true                                                                                                                                                                                                                    
--web.cert-file=/cert/tls.crt                                                                                                                                                                                                            
--web.key-file=/cert/tls.key                                                                                                                                                                                                             
--web.listen-address=:10250                                                                                                                                                                                                              
--web.tls-min-version=VersionTLS13
```

* Prometheus Operator Logs:

```
level=debug ts=2023-04-18T09:57:49.238252782Z caller=operator.go:611 component=prometheusagentoperator key=monitoring/test statefulset=prom-agent-test shard=0 msg="new statefulset generation inputs match current, skipping any actions"   
level=info ts=2023-04-18T09:57:49.238448063Z caller=operator.go:817 component=prometheusagentoperator key=monitoring/test msg="update prometheus status"                                                                                      
level=debug ts=2023-04-18T09:57:49.248057716Z caller=klog.go:84 component=k8s_client_runtime func=Infof msg="GET https://172.20.0.1:443/api/v1/namespaces/monitoring/pods?labelSelector=app.kubernetes.io%2Finstance%3Dtest%2Capp.kubernetes.i
level=debug ts=2023-04-18T09:57:49.307702761Z caller=klog.go:84 component=k8s_client_runtime func=Infof msg="PUT https://172.20.0.1:443/apis/monitoring.coreos.com/v1alpha1/namespaces/monitoring/prometheusagents/test/status 200 OK in 58 mi
level=warn ts=2023-04-18T09:57:49.30836959Z caller=klog.go:96 component=k8s_client_runtime func=Warning msg="unknown field \"spec.storage.ephemeral.volumeClaimTemplate.metadata.creationTimestamp\""                                          
panic: interface conversion: interface {} is *v1alpha1.PrometheusAgent, not *v1.Prometheus                                                                                                                                                     
                                                                                                                                                                                                                                               
goroutine 675 [running]:                                                                                                                                                                                                                       
github.com/prometheus-operator/prometheus-operator/pkg/prometheus.(*Collector).Collect(0xc0001a8b40, 0xc00056ff68?)                                                                                                                            
    /home/runner/work/prometheus-operator/prometheus-operator/pkg/prometheus/collector.go:69 +0xfc                                                                                                                                             
github.com/prometheus/client_golang/prometheus.(*wrappingCollector).Collect.func1()                                                                                                                                                            
    /home/runner/go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/wrap.go:131 +0x2e                                                                                                                                           
created by github.com/prometheus/client_golang/prometheus.(*wrappingCollector).Collect                                                                                                                                                         
    /home/runner/go/pkg/mod/github.com/prometheus/client_golang@v1.14.0/prometheus/wrap.go:130 +0x8d
```

**Comment by simonpasquier** [comment]
I confirm that it's a valid issue :)

The problem is that the Collect() method can get a PrometheusAgent pointer instead of Prometheus...

https://github.com/prometheus-operator/prometheus-operator/blob/27943fe43862d313a63b518b6ee67b59d7680c1d/pkg/prometheus/collector.go#L66-L72

We also need an additional label in the metrics to distinguish between agent and server resources.

cc @slashpai 
maybe @ArthurSens too if you have some cycles :)

**Comment by simonpasquier** [comment]
we also an additional label here
https://github.com/prometheus-operator/prometheus-operator/blob/27943fe43862d313a63b518b6ee67b59d7680c1d/pkg/operator/operator.go#L40-L45

**Comment by ArthurSens** [comment]
Oh, nice catch! Does the panic only happen when we try to scrape metrics from the operator?

**Comment by ArthurSens** [comment]
> We also need an additional label in the metrics to distinguish between agent and server resources.

Doesn't this line solve this?
https://github.com/prometheus-operator/prometheus-operator/blob/27943fe43862d313a63b518b6ee67b59d7680c1d/pkg/prometheus/agent/operator.go#L112

https://github.com/prometheus-operator/prometheus-operator/blob/27943fe43862d313a63b518b6ee67b59d7680c1d/pkg/prometheus/server/operator.go#L139

**Comment by CheyiLin** [comment]
> Oh, nice catch! Does the panic only happen when we try to scrape metrics from the operator?

There is an service-monitor created by helm for operator itself, so yes it does happen when scraping from the operator.

**Comment by simonpasquier** [comment]
Closed by #5511 and fixed in v0.64.1.
